### ISSUE TITLE
Assertion `status == napi_ok' failed.

### ISSUE THREAD

**Issue by hunterInt** [utropicmedia/storj-nodejs#11]
Randomly I will get this error when calling `upload_objectc`:

```
node: ../libUplink.cc:1857: napi_value__* upload_writec(napi_env, napi_callback_info): Assertion `status == napi_ok' failed.
Aborted (core dumped)
```

However it shuts down the NodeJS server when the error does happen. `catch`ing the error doesn't prevent the shutdown either. What does this error mean and any fix for this? 

Thanks!

**Comment by hunterInt** [comment]
Now it's happening every time I try to upload some data (`upload_objectc`) and thus not letting me upload any data. In the past, it happened once a week or two.

**Comment by shivam201312** [comment]
@hunterInt can you share code snippet ? 
we will try to replicate issue and get back to you .

**Comment by hunterInt** [comment]
@shivam201312 sure thing.

On app start up Storj is initializated:
```
/**
 * Initialization for Storj upload / download
 * 
 */

modules.storjInitialization = () => {
	return new Promise(async (resolve, reject) => {
		try {
			let storjObj = {};

			console.log("Getting Access\nSatellite Address : ", consts.storjConfig.satelliteURL);
			storjObj.accsessResult = await libUplink.request_access_with_passphrasec(consts.storjConfig.satelliteURL, consts.storjConfig.apiKey, consts.storjConfig.encryptionPassphrase);

			if (storjObj.accsessResult.error.message != '') {
				throw storjObj.accsessResult.error.message;
			}

			console.log('\nOpening the Storj project from satellite: ', consts.storjConfig.satelliteURL);
			storjObj.projectRef = await libUplink.open_projectc(storjObj.accsessResult.access);

			if (storjObj.projectRef.error.message != '') {
				throw storjObj.projectRef.error.message;
			}

			console.log('Creating restrict access....');

			const permission = new storj.Permission(true, true, true, true, 0, 0);
			let sharePrefix = storj.SharePrefix;
			let sharePrefixListArray = [];
			sharePrefix.bucket = consts.storjConfig.bucketName;
			sharePrefix.prefix = "";
			sharePrefixListArray.push(sharePrefix);
			storjObj.accessShareResult = libUplink.access_sharec(storjObj.accsessResult.access, permission, sharePrefixListArray, sharePrefixListArray.length);

			if ('' != storjObj.accessShareResult.error.message) {
				console.error('error');
				throw storjObj.accessShareResult.error.message;
			}

			console.log('Creating share access key...');
			let lO_StringResult = libUplink.access_serializec(storjObj.accessShareResult.access);

			if ('' != lO_StringResult.error.message) {
				console.error('error');
				throw lO_StringResult.error.message;
			}
			storjObj.accsessResult = libUplink.parse_accessc(lO_StringResult.string);
			console.log('Parsing Access created');
			console.log('Storj initialized');

			resolve(storjObj);

		} catch (error) {
			console.error('Storj initialization error: ', error);
			reject(false);
		}
	});

};
```

Error happens below at `upload_objectc`:
```
try {
	const filePath = true === isThumbnail ? username + '/thumbnail_' + fileUID : username + '/' + fileUID;

	console.log("Getting Upload Object....");

	let lO_uploadOptions = new storj.UploadOptions();
	lO_uploadOptions.expires = 0;

	storjObj.uploadObject = await libUplink.upload_objectc(storjObj.projectRef.project, consts.storjConfig.bucketName, filePath, lO_uploadOptions); //<----- error happens here

	if (storjObj.uploadObject.error.message != '') {
		console.log('Failed to get upload object');
		throw storjObj.uploadObject.error.message;
	}

...
```

This code worked perfectly for several weeks until yesterday for some reason.

**Comment by hunterInt** [comment]
@shivam201312 hello. Any luck?

**Comment by kmozurkewich** [comment]
@hunterInt  - we've just completed a migration of the repo to https://github.com/storj-thirdparty/uplink-nodejs along with an upgrade of the bindings to a newer version of libuplink.  

Can you try with the release at https://github.com/storj-thirdparty/uplink-nodejs and see if the issue still occurs?

**Comment by hunterInt** [comment]
Hi @kmozurkewich I tried cloning the https://github.com/storj-thirdparty/uplink-nodejs repo since I'm assuming running `npm install storj-nodejs` doesn't actually clone this new `storj-thirdparty`  repository. I then ran `go get storj.io/uplink-c` and then I tried starting my app. But I got this error:

```
/home/hunter/Documents/projects/sharp/node_modules/bindings/bindings.js:135
  throw err;
  ^

Error: Could not locate the bindings file. Tried:
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/Debug/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/Release/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/out/Debug/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/Debug/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/out/Release/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/Release/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/default/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/compiled/12.15.0/linux/x64/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/addon-build/release/install-root/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/addon-build/debug/install-root/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/addon-build/default/install-root/storjnodejs.node
 → /home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/lib/binding/node-v72-linux-x64/storjnodejs.node
    at bindings (/home/hunter/Documents/projects/sharp/node_modules/bindings/bindings.js:126:9)
    at Object.<anonymous> (/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/libUplinkNode.js:3:38)
    at Module._compile (internal/modules/cjs/loader.js:955:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:991:10)
    at Module.load (internal/modules/cjs/loader.js:811:32)
    at Function.Module._load (internal/modules/cjs/loader.js:723:14)
    at Module.require (internal/modules/cjs/loader.js:848:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at module.exports (/home/hunter/Documents/projects/sharp/app/fileObj.js:2:16)
    at Object.<anonymous> (/home/hunter/Documents/projects/sharp/server.js:15:41)
    at Module._compile (internal/modules/cjs/loader.js:955:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:991:10)
    at Module.load (internal/modules/cjs/loader.js:811:32)
    at Function.Module._load (internal/modules/cjs/loader.js:723:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1043:10)
    at internal/main/run_main_module.js:17:11 {
  tries: [
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/Debug/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/Release/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/out/Debug/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/Debug/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/out/Release/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/Release/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/build/default/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/compiled/12.15.0/linux/x64/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/addon-build/release/install-root/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/addon-build/debug/install-root/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/addon-build/default/install-root/storjnodejs.node',
    '/home/hunter/Documents/projects/sharp/node_modules/uplink-nodejs/lib/binding/node-v72-linux-x64/storjnodejs.node'
  ]
}

```

What do I need to run to get this to work?

**Comment by shivam201312** [comment]
@hunterInt after cloning  the project .Run following commands inside the cloned project
```
make
```
```
(node-gyp configure) && (node-gyp rebuild)
```

**Comment by hunterInt** [comment]
@shivam201312 thanks! The commands seem to have built the project correctly.

However, even after the update, I am still getting this error:

```
node: ../libUplink.cc:1857: napi_value__* upload_writec(napi_env, napi_callback_info): Assertion `status == napi_ok' failed.
Aborted (core dumped)

```

What do I need to do to fix this? Please advise.

**Comment by shivam201312** [comment]
@hunterInt  can you share code snippet of upload_writec function ?

**Comment by hunterInt** [comment]
@shivam201312 sure thing. Did you want the C++ `upload_write` function or the NodeJS usage of it in my application? The former is pasted here, and the latter is listed above in [this comment above.](https://github.com/utropicmedia/storj-nodejs/issues/11#issuecomment-632932049)

```
  //  * function upload_writec uploads len(p) bytes from p to the object's data stream.
  //  * pre-requisites: upload_objectc
  //  * inputs: Upload (Object) ,Pointer(Buffer) , Length (size_t)
  //  * output: WriteResult (Object)
napi_value upload_writec(napi_env env, napi_callback_info info) {
  napi_status status;
  
  size_t argc = 3;
  napi_value args[3];

  status = napi_get_cb_info(env, info,&argc,args, nullptr ,nullptr);
  assert(status == napi_ok);
  
  if (argc < 3) {
    napi_throw_type_error(env, nullptr, "\nWrong number of arguments!! excepted 3 arguments\n");
    return NULL;
  }

  napi_valuetype checktypeofinput;

  status = napi_typeof(env,args[0],&checktypeofinput);
  assert(status == napi_ok);
  if(checktypeofinput!=napi_object){
    napi_throw_type_error(env,nullptr,"\nWrong datatype !! first argument excepted to be object type\n");
    return NULL;
  }

  status = napi_typeof(env,args[1],&checktypeofinput);
  assert(status == napi_ok);
  if(checktypeofinput!=napi_object){
    napi_throw_type_error(env,nullptr,"\nWrong datatype !! second argument excepted to be object/buffer type\n");
    return NULL;
  }

  status = napi_typeof(env,args[2],&checktypeofinput);
  assert(status == napi_ok);
  if(checktypeofinput!=napi_number){
    napi_throw_type_error(env,nullptr,"\nWrong datatype !! third argument excepted to be number type\n");
    return NULL;
  }
  
  char *_err = "";
  char **err = &_err;
  bool propertyexists = false;
  napi_value ObjectkeyNAPI;
  char* handle = "_handle";
  status = napi_create_string_utf8(env,handle,NAPI_AUTO_LENGTH ,&ObjectkeyNAPI);
  assert(status == napi_ok);
  //
  status = napi_has_property(env,args[0],ObjectkeyNAPI,&propertyexists);
  assert(status == napi_ok);
  if(!propertyexists){
    napi_throw_type_error(env,nullptr,"\nInvalid Object \n");
    return NULL;
  }

  Upload upload_resultRef;
  upload_resultRef._handle = getHandleValue(env,args[0]);
  if(upload_resultRef._handle==0){
    return NULL;
  }
  bool buffertype1;
  status = napi_is_arraybuffer(env, args[1], &buffertype1);
  assert(status == napi_ok);
  size_t lengthOfBuffer;  

  void* bufferPtr = NULL;  

  status = napi_get_buffer_info(env, args[1], &bufferPtr,&lengthOfBuffer);
  assert(status == napi_ok);

  int64_t bytes64;
  status = napi_get_value_int64(env,args[2],&bytes64);
  assert(status == napi_ok);

  size_t bytesread = (size_t)bytes64;

  uint8_t *ptrToData ; 
  ptrToData = (uint8_t *)bufferPtr;

  WriteResult write_result = upload_write(&upload_resultRef,ptrToData,bytesread);
      napi_value returnObject,errorObject;
      status = napi_create_object(env,&returnObject);
      if(write_result.error!=NULL){
          Error error_result = *(write_result.error);
          char* errorMessagePtr = error_result.message;
          size_t datauploaded = 0;
          napi_value datauploadedObject;
          status = napi_create_object(env,&errorObject);
          assert(status == napi_ok);
          status = napi_create_int32(env,datauploaded,&datauploadedObject);
          assert(status == napi_ok);
          napi_value errorObject,errorCode,errorMessage;

          status = napi_set_named_property(env,returnObject,"bytes_written",datauploadedObject);
          assert(status == napi_ok);

          status = napi_create_int64(env, error_result.code, &errorCode);
          assert(status == napi_ok);

          status = napi_create_string_utf8(env,&errorMessagePtr[0],NAPI_AUTO_LENGTH,&errorMessage);
          assert(status == napi_ok);
          status = napi_set_named_property(env,errorObject,"code",errorCode);
          assert(status == napi_ok);
          
          status = napi_set_named_property(env,errorObject,"message",errorMessage);
          assert(status == napi_ok);

          status = napi_set_named_property(env,returnObject,"error",errorObject);
          assert(status == napi_ok);

          return returnObject;
      }else{
          char blank[] = "";
          napi_value errorObject,errorMessage;
          napi_value errorCode;
          status = napi_create_object(env,&errorObject);
          assert(status == napi_ok);

          size_t datauploaded = write_result.bytes_written;
          napi_value datauploadedObject;
          status = napi_create_int32(env,datauploaded,&datauploadedObject);
          assert(status == napi_ok);
          //
          status = napi_set_named_property(env,returnObject,"bytes_written",datauploadedObject);
          assert(status == napi_ok);

          status = napi_create_int64(env, 0, &errorCode);
          assert(status == napi_ok);

          status = napi_create_string_utf8(env,&blank[0],NAPI_AUTO_LENGTH,&errorMessage);
          assert(status == napi_ok);
          status = napi_set_named_property(env,errorObject,"code",errorCode);
          assert(status == napi_ok);
          
          status = napi_set_named_property(env,errorObject,"message",errorMessage);
          assert(status == napi_ok);

          status = napi_set_named_property(env,returnObject,"error",errorObject);
          assert(status == napi_ok);
          return returnObject;
      }
}

```

**Comment by shivam201312** [comment]
@hunterInt please share snippet of node js usage in your application.
I have use the snippet code that you have shared and just added file upload function . code is working fine for me . 

```
const fs = require("fs");

//include the Node.js-Storj bindings module
var storj = require("storj-nodejs");

const BUFFER_SIZE = 80000;
//
const RETRY_MAX = 5;
//
var libUplink = storj.storj_nodejs;
//
var storjConfig = {
    apiKey    : "change-me-to-the-api-key-created-in-satellite-gui",
    satelliteURL   : "us-central-1.tardigrade.io:7777",
    encryptionPassphrase  : "change-me-to-desired-bucket-name",
    bucketName   : "change-me-to-desired-bucket-name",
    uploadPath   : "optionalpath/requiredfilename",
};

// file names to be uploaded or downloaded as
var localFullFileName = {
    src : "change-me-to-source-file-name-at-local-system",
    dest: "change-me-to-destination-file-name-at-local-system",
};
//

//Storj Intialized
async function storjintialized(){
 return new Promise(async (resolve, reject) => {
  try {
   let storjObj = {};

   console.log("Getting Access\nSatellite Address : ", storjConfig.satelliteURL);
   storjObj.accsessResult = await libUplink.request_access_with_passphrasec(storjConfig.satelliteURL, storjConfig.apiKey, storjConfig.encryptionPassphrase);

   if (storjObj.accsessResult.error.message != '') {
    throw storjObj.accsessResult.error.message;
   }

   console.log('\nOpening the Storj project from satellite: ', storjConfig.satelliteURL);
   storjObj.projectRef = await libUplink.open_projectc(storjObj.accsessResult.access);

   if (storjObj.projectRef.error.message != '') {
    throw storjObj.projectRef.error.message;
   }

   console.log('Creating restrict access....');

   const permission = new storj.Permission(true, true, true, true, 0, 0);
   let sharePrefix = storj.SharePrefix;
   let sharePrefixListArray = [];
   sharePrefix.bucket = storjConfig.bucketName;
   sharePrefix.prefix = "";
   sharePrefixListArray.push(sharePrefix);
   storjObj.accessShareResult = libUplink.access_sharec(storjObj.accsessResult.access, permission, sharePrefixListArray, sharePrefixListArray.length);

   if ('' != storjObj.accessShareResult.error.message) {
    console.error('error');
    throw storjObj.accessShareResult.error.message;
   }

   console.log('Creating share access key...');
   let lO_StringResult = libUplink.access_serializec(storjObj.accessShareResult.access);

   if ('' != lO_StringResult.error.message) {
    console.error('error');
    throw lO_StringResult.error.message;
   }
   storjObj.accsessResult = libUplink.parse_accessc(lO_StringResult.string);
   console.log('Parsing Access created');
   console.log('Storj initialized');

   resolve(storjObj);

  } catch (error) {
   console.error('Storj initialization error: ', error);
   reject(false);
  }
 });
}

//
storjintialized().then((storjObj)=>{
 console.log(storjObj);
 //const filePath = true === isThumbnail ? username + '/thumbnail_' + fileUID : username + '/' + fileUID;

 let lO_uploadOptions = new storj.UploadOptions();
 lO_uploadOptions.expires = 0;
    //
    
 storjObj.uploadObject = libUplink.upload_objectc(storjObj.projectRef.project,storjConfig.bucketName, storjConfig.uploadPath, lO_uploadOptions); //<----- error happens here
 console.log(storjObj.uploadObject);
 if (storjObj.uploadObject.error.message != '') {
  console.log('Failed to get upload object');
  throw storjObj.uploadObject.error.message;
 }

 console.log(localFullFileName.src, " File: UPLOADED as ", storjConfig.uploadPath, "!");
            
    var fileHandle = fs.openSync(localFullFileName.src, "rs");
    
    var size = {
                    file            : `${fs.statSync(localFullFileName.src).size}`,
                    toWrite         : 0,
                    actuallyWritten : 0,
                    totalWritten    : 0
                };
    var li_retryCount = RETRY_MAX;
                    
    var buffer = new Buffer.alloc(BUFFER_SIZE);
    //
    while (size.totalWritten < size.file) { 
                size.toWrite = size.file - size.totalWritten;
                
                if (size.toWrite > BUFFER_SIZE) {
                    size.toWrite = BUFFER_SIZE;
                } else if (size.toWrite == 0) {
                    break;
                }

                // read proper number of bytes from the file
                var bytesRead = fs.readSync(fileHandle, buffer, 0, size.toWrite, size.totalWritten);
                
                while ((bytesRead <= 0) && (--li_retryCount > 0)) {
                    //
                    bytesRead = fs.readSync(fileHandle, buffer, 0, size.toWrite, size.totalWritten);
                }
                //
                if (bytesRead <= 0) {
                    // could not read further
                    ls_error = "ERROR: Read " + size.totalWritten.toString() + " bytes of complete " + size.file.toString() + " bytes from the file. Could not read further.";
                    break;
                }
                
                if (li_retryCount < RETRY_MAX) {
                    li_retryCount = RETRY_MAX;
                }

                var ls_error = "";               
                //
                do {
                    // upload the buffer's content to the Storj bucket bytes_written
                    var uploadedInfo = libUplink.upload_writec(storjObj.uploadObject.upload, buffer, bytesRead);
                    if (uploadedInfo.error.message != "") {
                        ls_error = uploadedInfo.error.message;
                    } else {
                        size.actuallyWritten = uploadedInfo.bytes_written;
                        if(size.actuallyWritten>=size.file){
                            break;
                        }
                    }
                } while ((ls_error != "") && (--li_retryCount > 0));
                
                if (ls_error != "") {
                    break;
                }
                
                if (li_retryCount < RETRY_MAX) {
                    li_retryCount = RETRY_MAX;
                }
                
                // if no further data is left to be read from the file and then uploaded
                if (size.actuallyWritten == 0) {
                    break;
                }
                //
                size.totalWritten += size.actuallyWritten;
                if((size.totalWritten>0)&&(size.file>0)){
                    console.log("File uploaded on storj  : ",((Number(size.totalWritten)/Number(size.file))*100).toFixed(4)," %");
                }
                
            }
                
            fs.closeSync(fileHandle);
      console.log("Commiting Object on storj...");
            var uploadcommitResult = libUplink.upload_commitc(storjObj.uploadObject.upload);
            console.log(uploadcommitResult);
            if(uploadcommitResult.error.message==""){
                console.log("Object Uploaded on storj : ",storjConfig.uploadPath);
            }else{
                console.log("FAILed to upload object on storj");
                
            }
    //
}).catch((error)=>{
 console.log("Error");
 console.log(error);
})
```

**Comment by hunterInt** [comment]
@shivam201312 oops my bad, I thought I pasted in my `upload_writec` usage above. Here it is: 

```
try {
	const filePath = true === isThumbnail ? username + '/thumbnail_' + fileUID : username + '/' + fileUID;

	console.log("Getting Upload Object....");

	let lO_uploadOptions = new storj.UploadOptions();
	lO_uploadOptions.expires = 0;

	storjObj.uploadObject = await libUplink.upload_objectc(storjObj.projectRef.project, consts.storjConfig.bucketName, filePath, lO_uploadOptions);

	if (storjObj.uploadObject.error.message != '') {
		console.log('Failed to get upload object');
		throw storjObj.uploadObject.error.message;
	}

	//Length of the total encrypted file string
	const encryptedFileStringByteLength = Buffer.byteLength(fileEncryptedString, 'utf8');

	//The size of the first chunk of the fileEncryptedString
	let indexStart = 0;
	let indexEnd = bufferSize;

	let encryptedFileStringShort = '';
	let totalWritten = 0;

	//Loop until all uploaded
	while (totalWritten < encryptedFileStringByteLength) {
		encryptedFileStringShort = fileEncryptedString.substring(indexStart, indexEnd);

		const buffer = Buffer.from(encryptedFileStringShort);
		const bufferLength = buffer.length;

		let uploadError = '';
		let retryCount = retryMax;

		//While we're below the max retry count, and if we have errors, keep trying
		do {
			//Upload the buffer's content to the Storj bucket
			storjObj.uploadedInfo = await libUplink.upload_writec(storjObj.uploadObject.upload, buffer, bufferLength);
			if (storjObj.uploadedInfo.error.message != '') {
				uploadError = storjObj.uploadedInfo.error.message;
			} else {
				totalWritten = totalWritten + storjObj.uploadedInfo.bytes_written;

				//Reset uploadError
				uploadError = '';
			}

		} while ((uploadError != '') && (--retryCount > 0));

		//If there is still an error even after retrying several times
		if (uploadError != '') {
			throw uploadError;
		}

		//console.log("Data written on storj : ", totalWritten);

		//Shift to the next section of the fileEncryptedString
		indexStart = indexEnd;
		indexEnd = totalWritten + bufferSize;
	}

	if (encryptedFileStringByteLength !== totalWritten) {
		throw 'Whole file not written';
	}

	storjObj.uploadcommit = await libUplink.upload_commitc(storjObj.uploadObject.upload);

	if (storjObj.uploadcommit.error.message != '') {
		console.log('Failed to upload object on storj');
		throw storjObj.uploadcommit.error.message;
	}
	console.log('Object uploaded on storj');

```

**Comment by shivam201312** [comment]
Hi.. @hunterInt , this code is also working fine . i am able to upload data on storj V3 .
```
const fs = require("fs");

//include the Node.js-Storj bindings module
var storj = require("storj-nodejs");
var randomstring = require("randomstring");
//
//const BUFFER_SIZE = 80000;
const BUFFER_SIZE = 8;
//
const RETRY_MAX = 5;
//
var libUplink = storj.storj_nodejs;
//
var storjConfig = {
    apiKey    : "change-me-to-desried-API-Key",
    satelliteURL   : "us-central-1.tardigrade.io:7777",
    encryptionPassphrase  : "test",
    bucketName   : "change-me-to-desired-bucket-name",
    uploadPath   : "filepath/uploadobject.txt",
};

//Storj Intialized
async function storjintialized(){
	return new Promise(async (resolve, reject) => {
		try {
			let storjObj = {};

			console.log("Getting Access\nSatellite Address : ", storjConfig.satelliteURL);
			storjObj.accsessResult = await libUplink.request_access_with_passphrasec(storjConfig.satelliteURL, storjConfig.apiKey, storjConfig.encryptionPassphrase);

			if (storjObj.accsessResult.error.message != '') {
				throw storjObj.accsessResult.error.message;
			}

			console.log('\nOpening the Storj project from satellite: ', storjConfig.satelliteURL);
			storjObj.projectRef = await libUplink.open_projectc(storjObj.accsessResult.access);

			if (storjObj.projectRef.error.message != '') {
				throw storjObj.projectRef.error.message;
			}

			console.log('Creating restrict access....');

			const permission = new storj.Permission(true, true, true, true, 0, 0);
			let sharePrefix = storj.SharePrefix;
			let sharePrefixListArray = [];
			sharePrefix.bucket = storjConfig.bucketName;
			sharePrefix.prefix = "";
			sharePrefixListArray.push(sharePrefix);
			storjObj.accessShareResult = libUplink.access_sharec(storjObj.accsessResult.access, permission, sharePrefixListArray, sharePrefixListArray.length);

			if ('' != storjObj.accessShareResult.error.message) {
				console.error('error');
				throw storjObj.accessShareResult.error.message;
			}

			console.log('Creating share access key...');
			let lO_StringResult = libUplink.access_serializec(storjObj.accessShareResult.access);

			if ('' != lO_StringResult.error.message) {
				console.error('error');
				throw lO_StringResult.error.message;
			}
			storjObj.accsessResult = libUplink.parse_accessc(lO_StringResult.string);
			console.log('Parsing Access created');
			console.log('Storj initialized');

			resolve(storjObj);

		} catch (error) {
			console.error('Storj initialization error: ', error);
			reject(false);
		}
	});
}

//
storjintialized().then(async (storjObj)=>{
	//const filePath = true === isThumbnail ? username + '/thumbnail_' + fileUID : username + '/' + fileUID;
    //
	let lO_uploadOptions = new storj.UploadOptions();
	lO_uploadOptions.expires = 0;
    //
	storjObj.uploadObject = await libUplink.upload_objectc(storjObj.projectRef.project,storjConfig.bucketName, storjConfig.uploadPath, null); //<----- error happens here
	if (storjObj.uploadObject.error.message != '') {
		console.log('Failed to get upload object');
		throw storjObj.uploadObject.error.message;
	}
    //
    var li_retryCount = RETRY_MAX;
    //
    fileEncryptedString = randomstring.generate(1000);
    //
    var buffer = new Buffer.alloc(BUFFER_SIZE);
    //
   	//Length of the total encrypted file string
    const encryptedFileStringByteLength = Buffer.byteLength(fileEncryptedString, 'utf8');
    //The size of the first chunk of the fileEncryptedString
    let indexStart = 0;
    let indexEnd = BUFFER_SIZE;
    //
    let encryptedFileStringShort = "";
    let totalWritten = 0;
    //Loop until all uploaded
    while (totalWritten < encryptedFileStringByteLength) {
        encryptedFileStringShort = fileEncryptedString.substring(indexStart, indexEnd);
        //
        const buffer = Buffer.from(encryptedFileStringShort);
        const bufferLength = buffer.length;
        let uploadError = '';
        let retryCount = RETRY_MAX;
        //While we're below the max retry count, and if we have errors, keep trying
        do {
            //Upload the buffer's content to the Storj bucket
            storjObj.uploadedInfo =await libUplink.upload_writec(storjObj.uploadObject.upload, buffer, bufferLength);
            if (storjObj.uploadedInfo.error.message != '') {
                uploadError = storjObj.uploadedInfo.error.message;
            } else {
                totalWritten = totalWritten + storjObj.uploadedInfo.bytes_written;
                //Reset uploadError
                uploadError = '';
            }
        } while ((uploadError != '') && (--retryCount > 0));
        //If there is still an error even after retrying several times
        if (uploadError != '') {
            throw uploadError;
        }
        console.log("Data written on storj : ", totalWritten);

        //Shift to the next section of the fileEncryptedString
        indexStart = indexEnd;
        indexEnd = totalWritten + bufferLength;
    }

    if (encryptedFileStringByteLength !== totalWritten) {
        throw 'Whole file not written';
    }
    storjObj.uploadcommit =await libUplink.upload_commitc(storjObj.uploadObject.upload);
    if (storjObj.uploadcommit.error.message != '') {
        console.log('Failed to upload object on storj');
        throw storjObj.uploadcommit.error.message;
    }
    console.log('Object uploaded on storj');
    //
}).catch((error)=>{
	console.log("Error");
	console.log(error);
})
```

**Comment by kaloyan-raev** [comment]
@hunterInt As far as I understand, the library crashes for you at this exact place: https://github.com/storj-thirdparty/uplink-nodejs/blob/master/libUplink.cc#L1857

This is the place where the upload failed and the library is constructing the response with error details. Perhaps, there is something wrong with how the library calls the N-API and it causes the crashes.

Something that would help to analyze the issue is if you could print the value of a few variables just before this problematic line (i.e. between lines 1856 and 1857):
- `error_result.code`
- `error_result.message`
- `status`

Could you add these print statements, rebuild the library, and reproduce the issue?

@shivam201312 It would be worth having an automated test that simulates upload error, so we are sure this part of the code is robust and does not crash.

**Comment by hunterInt** [comment]
Thanks for taking a look @kaloyan-raev! I noticed that the link you sent was referencing a newer version of the nodeJS uplink library so I went ahead and updated. This seemed to fix the ```
Assertion `status == napi_ok' failed.``` issue since I am no longer getting that error. 

However, the app now locks up on `upload_commit`:

```
storjObj.uploadcommit = await libUplink.upload_commit(storjObj.uploadObject.upload);
```

And requires a system reboot or for the process to be killed.
